### PR TITLE
anim_goal_use_object and other additions

### PR DESCRIPTION
**Pull request by anatoliy-savchak**
* action_sequence.cpp - fixed dagger not thrown by ai;
* ai.cpp - added "use item", and "d20 action", added more logging info; moved tactic middle string parsing to StrategyTabLineParseTacticMiddleString;
* mapsystem_hooks.cpp - fixed crash in CheckFogForCritter@0x100327A0;
* history.cpp\.h - exposed RollHistoryType2Add;
* inventory.cpp - fixed dagger not thrown by ai; exposed UiOpenContainer;
* python_dispatcher.cpp - added bonus_list to EventObjD20Action;
* python_game.cpp - added char_ui_display_type;
* python_object.cpp - added objhndl.anim_goal_use_object, objhndl.use_item and objhndl.container_open_ui;

### CODE REVIEW

**Comment by anatoliy-savchak** [comment]
Release notes:
* added ability for AI to throw daggers;
* added PyObjHandle.**anim_goal_use_object**(PyObjHandle: target, int: goal_type = 40, int: target_loc = None, int: some_flag = None) -> int
* added PyObjHandle.**use_item**(PyObjHandle: item, PyObjHandle: target = None) -> int
* added PyObjHandle.**container_open_ui**(PyObjHandle: container) -> int
* added tpdp.EventObjD20Action.**bonus_list**;
* added tactic "**use item**" with args str: id;
* added tactic "**d20 action**" with args int: d20action, int: data1;
* added game.**char_ui_display_type**;

File: TemplePlus/python/python_object.cpp [hunk 1]
```diff
@@ -3280,6 +3311,77 @@ static PyObject* PyObjHandle_Unconceal(PyObject* obj, PyObject* args) {
...
+	}
+	d20Sys.GlobD20ActnInit();
+	d20Sys.GlobD20ActnSetTypeAndData1(D20A_USE_POTION, 0);
+	actSeqSys.ActSeqCurSetSpellPacket(&spellPktBody, 1);
+	D20SpellData d20SpellData;
+	d20Sys.D20ActnSetSpellData(&d20SpellData, spellPktBody.spellEnum, spellPktBody.spellClass, spellPktBody.casterLevel, 0xFF, 0);
+	d20Sys.GlobD20ActnSetSpellData(&d20SpellData);
+	d20Sys.GlobD20ActnSetTarget(target, &loc);
+	ActionErrorCode result = (ActionErrorCode)actSeqSys.ActionAddToSeq();
+	while (!result) {
```
Comment (doug1234) [inline TemplePlus/python/python_object.cpp:3374]: This while loop looks like it exits the first time through every time.  Could this be rewritten with if statements.  I am sure you tested the code but this is very strange.

**Comment by doug1234** [comment]
Any chance you could change the while loop in PyObjHandle_UseItem() which is very strange.  Other than that the changes look good.

**Comment by DudeMcDude** [comment]
Can you elaborate the crash in the CheckFogForCritter@0x100327A0? It sounds a little suspicious, as if you're treating the symptom and not the problem.

**Comment by anatoliy-savchak** [comment]
> Any chance you could change the while loop in PyObjHandle_UseItem() which is very strange. Other than that the changes look good and I am ready to merge it.

I tried to find proper name for this control flow technique, but not found any. Old-school guys from my previous job tech me that. The idea is following:
```
while(1):
	if (condition1):
		do stuff 1
		break

	if (condition2):
		arg = heavy_calculation()
		if (arg == 1):
			more stuff
		elif (arg == 1):
			more stuff 2
		elif (arg == 3):
			update_local_vars()
			continue #restart check
		break

	default stuff
	break
```
But that can be confusing, I will remove it.

There will be additional changes to the use_item code. Along with rewrite of tactics.use_item to use this code, as old one does not work sometimes.

Plus, I found bug in D20ActionCallbacks::PerformCastItemSpell, the check of identified should be for PC only.
`if (casterObj->type == obj_t_pc && !inventory.IsIdentified(item) && itemObj->type != obj_t_food){ // blind use of magic item, for pc only`

In the recent video I showed how NPC used wand. Without the fix it would not.

**Comment by anatoliy-savchak** [comment]
> Can you elaborate the crash in the CheckFogForCritter@0x100327A0? It sounds a little suspicious, as if you're treating the symptom and not the problem.

Yes, I'm trying to fix a symptom. I encountered it a while ago, have not found obvious reason for the bug. I can add log info or remove it.

**Comment by anatoliy-savchak** [comment]
I would like to move code from AiSystem::UseItem and PyObjHandle_UseItem to common place, as it is essentially the same. Where should I put it? combat.h maybe?

**Comment by anatoliy-savchak** [comment]
Also, how can I print log using actionErrorCodeStrings?

**Comment by doug1234** [comment]
Thanks for making the change!  I haven't seen code like that before.  

Could you call the code in the AI file from the python file?  That would seem like a reasonable way to do it to me.

**Comment by anatoliy-savchak** [comment]
I think this should be final commit on this matter. tested extensively this use item. Now potion, wand and scrolls are all tested.

**Comment by doug1234** [comment]
I am pretty satisfied at this point.  I am going to go ahead and merge it for releasing a new version soon.

**Pull request merged**